File: backend/src/service/token.service.ts
import { getRedisClient, CacheKeys, CacheTTL } from '@/config/redis'
import type { userInfo } from '@/utils/auth'

/**
 * Token 缓存数据结构
 */
interface TokenCacheData {
    accessToken: string
    userId: number
    username: string
    cover: string | null
    signature: string | null
    createdAt: number // 存入Redis的时间戳
}

/**
 * 将用户token存入Redis
 * @param userId 用户ID
 * @param accessToken 访问令牌
 * @param userInfo 用户信息
 */
export async function saveTokenToRedis(
    userId: number,
    accessToken: string,
    userInfo: {
        username: string
        cover: string | null
        signature: string | null
    }
): Promise<void> {
    try {
        const redis = getRedisClient()
        const cacheKey = CacheKeys.userToken(userId)

        const tokenData: TokenCacheData = {
            accessToken,
            userId,
            username: userInfo.username,
            cover: userInfo.cover,
            signature: userInfo.signature,
            createdAt: Date.now()
        }

        // 存入Redis,设置过期时间为2小时
        await redis.setex(
            cacheKey,
            CacheTTL.USER_TOKEN,
            JSON.stringify(tokenData)
        )

        console.log(`Token已存入Redis, userId: ${userId}`)
    } catch (error) {
        console.error('Token存入Redis失败:', error)
        // 不抛出错误,避免影响登录流程
    }
}

/**
 * 从Redis获取用户token
 * @param userId 用户ID
 * @returns Token数据或null
 */
export async function getTokenFromRedis(userId: number): Promise<TokenCacheData | null> {
    try {
        const redis = getRedisClient()
        const cacheKey = CacheKeys.userToken(userId)

        const cachedData = await redis.get(cacheKey)

        if (!cachedData) {
            return null
        }

        const tokenData: TokenCacheData = JSON.parse(cachedData)

        // 检查是否过期（双重保险）
        const elapsedSeconds = (Date.now() - tokenData.createdAt) / 1000
        if (elapsedSeconds > CacheTTL.USER_TOKEN) {
            // 已过期,删除缓存
            await redis.del(cacheKey)
            return null
        }

        console.log(`从Redis获取Token成功, userId: ${userId}`)
        return tokenData
    } catch (error) {
        console.error('从Redis获取Token失败:', error)
        return null
    }
}

/**
 * 删除Redis中的用户token
 * @param userId 用户ID
 */
export async function deleteTokenFromRedis(userId: number): Promise<void> {
    try {
        const redis = getRedisClient()
        const cacheKey = CacheKeys.userToken(userId)
        await redis.del(cacheKey)
        console.log(`Token已从Redis删除, userId: ${userId}`)
    } catch (error) {
        console.error('Token删除失败:', error)
    }
}

File: backend/java-backend/src/main/java/com/example/demo/service/TokenService.java
package com.example.demo.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.util.concurrent.TimeUnit;

@Slf4j
@Service
public class TokenService {

    private static final String TOKEN_KEY_PREFIX = "user:token:";
    private static final long TOKEN_TTL_HOURS = 2;

    private final StringRedisTemplate redisTemplate;
    private final ObjectMapper objectMapper;

    public TokenService(StringRedisTemplate redisTemplate, ObjectMapper objectMapper) {
        this.redisTemplate = redisTemplate;
        this.objectMapper = objectMapper;
    }

    /** 将用户Token和基础信息缓存到Redis，2小时过期 */
    public void saveTokenToRedis(Integer userId, String accessToken, Object userInfo) {
        try {
            TokenCacheData cacheData = new TokenCacheData();
            cacheData.setAccessToken(accessToken);
            cacheData.setUserId(userId);
            cacheData.setCreatedAt(System.currentTimeMillis());

            // 从 userInfo 对象提取字段
            if (userInfo instanceof com.example.demo.common.JwtUtil.UserInfo info) {
                cacheData.setUsername(info.getUsername());
                cacheData.setCover(info.getCover());
                cacheData.setSignature(info.getSignature());
            }

            String json = objectMapper.writeValueAsString(cacheData);
            String key = TOKEN_KEY_PREFIX + userId;
            redisTemplate.opsForValue().set(key, json, TOKEN_TTL_HOURS, TimeUnit.HOURS);
            log.debug("Token saved to Redis for userId: {}", userId);
        } catch (JsonProcessingException e) {
            log.error("Failed to save token to Redis for userId: {}", userId, e);
        }
    }

    /** 从Redis获取用户Token信息，过期则删除并返回null */
    public TokenCacheData getTokenFromRedis(Integer userId) {
        try {
            String key = TOKEN_KEY_PREFIX + userId;
            String json = redisTemplate.opsForValue().get(key);
            if (json == null) {
                return null;
            }

            TokenCacheData cacheData = objectMapper.readValue(json, TokenCacheData.class);

            // 检查是否过期（2小时）
            long elapsed = System.currentTimeMillis() - cacheData.getCreatedAt();
            if (elapsed > TOKEN_TTL_HOURS * 60 * 60 * 1000) {
                deleteTokenFromRedis(userId);
                return null;
            }

            return cacheData;
        } catch (JsonProcessingException e) {
            log.error("Failed to get token from Redis for userId: {}", userId, e);
            return null;
        }
    }

    /** 从Redis删除指定用户的Token缓存 */
    public void deleteTokenFromRedis(Integer userId) {
        String key = TOKEN_KEY_PREFIX + userId;
        redisTemplate.delete(key);
        log.debug("Token deleted from Redis for userId: {}", userId);
    }

    @Data
    public static class TokenCacheData {
        private String accessToken;
        private Integer userId;
        private String username;
        private String cover;
        private String signature;
        private long createdAt;
    }
}
